lib/Conversion/TritonGPUToLLVM/TensorMemRefOpToLLVM.cpp

```diff
@@ -0,0 +1,120 @@
+
+  TensorToMemRefOpConversion(LLVMTypeConverter &converter,
+                             AxisInfoAnalysis &axisAnalysisPass, 
+                             PatternBenefit benefit)
+      : ConvertTritonGPUOpToLLVMPattern<triton::gpu::TensorToMemRefOp>(converter, benefit) {}
+  LogicalResult 
+  matchAndRewrite(triton::gpu::TensorToMemRefOp op, 
+                  OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    Value src = op.getSrc();
+    auto srcTy = src.getType().dyn_cast<RankedTensorType>();
+    Attribute srcEncoding = srcTy.getEncoding();
+    if (!(srcEncoding && srcEncoding.isa<LDSEncodingAttr>())) return failure();
+
+    Value smemBase = getSharedMemoryBase(loc, rewriter, src);
+
+    Value res = op.getResult();
+    auto resTy = res.getType().dyn_cast<MemRefType>();
+
+    MemRefDescriptor desc = MemRefDescriptor::fromStaticShape(rewriter, loc,
+		                                              *getTypeConverter(), resTy,
+					                      smemBase, 
+							      smemBase);
+    rewriter.replaceOp(op, {desc});
+    return success();
+  }
+};
+
+struct MemRefToTensorOpConversion 
+    : public ConvertTritonGPUOpToLLVMPattern<triton::gpu::MemRefToTensorOp> {
+  using ConvertTritonGPUOpToLLVMPattern<
+      triton::gpu::MemRefToTensorOp>::ConvertTritonGPUOpToLLVMPattern;
+  MemRefToTensorOpConversion(LLVMTypeConverter &converter,
+                             AxisInfoAnalysis &axisAnalysisPass, 
+                             PatternBenefit benefit)
+      : ConvertTritonGPUOpToLLVMPattern<triton::gpu::MemRefToTensorOp>(converter, benefit) {}
+  LogicalResult 
+  matchAndRewrite(triton::gpu::MemRefToTensorOp op, 
+                  OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    Value src = op.getSrc();
+    auto srcTy = src.getType().dyn_cast<MemRefType>();
+    auto srcRank = srcTy.getRank();
+    auto srcShape = srcTy.getShape();
+    auto srcElemTy = srcTy.getElementType().dyn_cast<VectorType>();
+    Type llSrcElemTy = typeConverter->convertType(srcElemTy);
+    auto srcElemSize  = srcElemTy.getNumElements();
+
+    auto gpuAttr = srcTy.getMemorySpace().dyn_cast<mlir::gpu::AddressSpaceAttr>();
+    if (gpuAttr.getValue() != mlir::gpu::GPUDialect::getPrivateAddressSpace()) {
+      llvm::errs() << "private address space is required.\n";
+      return failure();
+    }
+    
+    Value res = op.getResult();
+    auto resTy = res.getType().dyn_cast<RankedTensorType>();
+    Type resElemTy = typeConverter->convertType(resTy.getElementType());
+    auto resRank = resTy.getRank();
+
+    if (( resRank == 0) || (srcRank == 0)) {
+      llvm::errs() << "Non-scalar rank is required\n";
+      return failure();
+    }
+
+    unsigned numElems = getElemsPerThread(resTy);
+    auto srcNumElems = product<long>(srcShape);
```

Comment:
> You also need to consider the `vector` element type to get the correct number of total elements inside memref.

The number of vectors in memref is srcNumElems. srcShape is the shape of memref. line 87 gives you the number of vectors in the memref.

Reply:
So you also need to get the number of elements in the vector so that you can have the total number of elements in the memref, which should match `numElems`.

Reply:
> So you also need to get the number of elements in the vector so that you can have the total number of elements in the memref, which should match `numElems`.

The number of elements in a vector is srcElemSize. I could not find an implementation of getElemsPerThread for mfma encoding. I thought getElemsPerThread would return the repetitions of tensor core mma computation. Have you implemented it somewhere?

Reply:
If this is the src
```
memref<4xvector<16xf32>>
```
Then srcShape would be [4], srcElemType would be `vector<16xf32>`, so srcElemSize is 4.

Mfma::getElemsPerThread() is not implemented yet. It'll be done soon. If it is blocking you, you can simply get the shape of the tensor and do `product<>()` to get the number of elements.